client/rpc/src/eth.rs

```diff
@@ -47,6 +47,123 @@ use crate::{internal_err, error_on_execution_failure, EthSigner};
 pub use fc_rpc_core::{EthApiServer, NetApiServer, Web3ApiServer};
 use codec::{self, Encode};
 
+mod estimate_gas_binary {
+	use ethereum_types::{H160, H256, U256};
+	use pallet_evm::{
+		BackendT, EvmBasic, Vicinity, Config, StackExecutor,
+		Precompiles as PrecompilesT, ExitReason
+	};
+	use fc_rpc_core::types::CallRequest;
+
+	pub struct Backend<'vicinity> {
+		_vicinity: &'vicinity Vicinity,
+	}
+
+	impl<'vicinity> Backend<'vicinity> {
+		pub fn new(_vicinity: &'vicinity Vicinity) -> Self {
+			Self { _vicinity }
+		}
+	}
+
+	impl<'vicinity> BackendT for Backend<'vicinity> {
+		fn gas_price(&self) -> U256 { U256::zero() }
+		fn origin(&self) -> H160 { H160::default() }
+		fn block_hash(&self, _number: U256) -> H256 { H256::default() }
+		fn block_number(&self) -> U256 { U256::zero() }
+		fn block_coinbase(&self) -> H160 { H160::default() }
+		fn block_timestamp(&self) -> U256 { U256::zero() }
+		fn block_difficulty(&self) -> U256 { U256::zero() }
+		fn block_gas_limit(&self) -> U256 { U256::zero() }
+		fn chain_id(&self) -> U256 { U256::zero() }
+		fn exists(&self, _address: H160) -> bool { true }
+		fn basic(&self, _address: H160) -> EvmBasic {
+			EvmBasic {
+				balance: U256::zero(),
+				nonce: U256::zero(),
+			}
+		}
+		fn code_size(&self, _address: H160) -> usize { 0 as usize }
+		fn code_hash(&self, _address: H160) -> H256 { H256::default() }
+		fn code(&self, _address: H160) -> Vec<u8> { Vec::new() }
+		fn storage(&self, _address: H160, _index: H256) -> H256 { H256::default() }
+	}
+
+	pub fn execute(request: CallRequest) -> (ExitReason, U256) {
+
+		type Precompiles = (
+			pallet_evm::precompiles::ECRecover,
+			pallet_evm::precompiles::Sha256,
+			pallet_evm::precompiles::Ripemd160,
+			pallet_evm::precompiles::Identity,
+		);
+
+		let CallRequest {
+			from,
+			to,
+			gas_price,
+			gas,
+			value,
+			data,
+			..
+		} = request;
+
+		let evm_config = Config::istanbul();
+
+		let gas_limit = gas.unwrap_or(U256::max_value()); // TODO: set a limit
+
+		let vicinity = Vicinity {
+			gas_price: gas_price.unwrap_or_default(),
+			origin: to.unwrap_or_default(),
+		};
+
+		let backend = Backend::new(&vicinity);
+
+		let mut high: u32 = gas_limit.low_u32();
+		let mut low: u32 = 21000;
+
+		let mut current: u32;
+		let mut exit_reason: Option<ExitReason> = None;
+		while low + 1 < high {
+			current = (high.saturating_add(low)) / 2;
+
+			let mut executor = StackExecutor::new_with_precompile(
+				&backend,
+				current as usize,
+				&evm_config,
+				<Precompiles as PrecompilesT>::execute
```

Comment:
@sorpaas will the function pointer work defining the type locally and infering it like this?